subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java

```diff
@@ -106,7 +103,12 @@ public JavaCompile() {
         ObjectFactory objectFactory = getObjectFactory();
         compileOptions = objectFactory.newInstance(CompileOptions.class);
         modularity = objectFactory.newInstance(DefaultModularitySpec.class);
-        javaCompiler = objectFactory.property(JavaCompiler.class);
+        JavaToolchainService javaToolchainService = getJavaToolchainService();
+        Provider<JavaCompiler> javaCompilerConvention = getProviderFactory()
+            .provider(() -> JavaCompileExecutableUtils.getExecutableOverrideToolchainSpec(this, objectFactory))
+            .orElse(objectFactory.newInstance(DefaultToolchainSpec.class))
+            .flatMap(javaToolchainService::compilerFor);
```

Comment:
```suggestion
        Provider<JavaCompiler> javaCompilerConvention = getProviderFactory()
            .provider(() -> JavaCompileExecutableUtils.getExecutableOverrideToolchainSpec(this, objectFactory))
            .flatMap(javaToolchainService::compilerFor)
            .orElse(javaToolchainService.compilerFor(it -> {}));
```

Would this be a working alternative without having to resort to `DefaultToolchainSpec`?

Reply:
This seems to be a good practical solution. Good idea!

Although, I still feel it is not as readable as `.orElse(Fallback...` or `.orElse(Default...)`, but having fewer internal dependencies in tasks implementations should be a long-term goal.

I applied this small change to all the Java tasks.